{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from pathlib import Path\n",
    "import fuc\n",
    "from fuc import pyvcf\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set path for the ouputs of the cyp2d6 caller\n",
    "caller_outputs_path = Path(\"/home/jupyter-yusuf/aws-s3-cyp2d6/cyp2d6-caller-outputs\") \n",
    "\n",
    "# Load list of samples with potential novel alleles\n",
    "\n",
    "file_path = \"samples_with_potential_novel_alleles.txt\"\n",
    "\n",
    "with open(\"samples_with_potential_novel_alleles.txt\", \"r\") as f:\n",
    "            samples = [line.strip() for line in f.readlines() if len(line) > 0]\n",
    "sample_list = [name for name in samples]\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Function to parse stellar pgx output and to obtain all the merged vcfs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_output_stellarpgx(stellarpgx_alleles_file):\n",
    "    file_path = Path(stellarpgx_alleles_file).resolve()\n",
    "    if not file_path.exists():\n",
    "        raise FileNotFoundError(file_path)\n",
    "\n",
    "    with file_path.open(\"r\") as f:\n",
    "        lines = [line.strip() for line in f.readlines() if len(line) > 0]\n",
    "\n",
    "    result = []\n",
    "    for i, line in enumerate(lines):\n",
    "        if line.startswith(\"Likely\"):\n",
    "            result.append(lines[i + 1])\n",
    "            \n",
    "        elif line.startswith(\"Initially\"):\n",
    "            result.append(lines[i][-1])\n",
    "            \n",
    "        elif line.startswith(\"Candidate\"):\n",
    "            result.append(lines[i + 1].strip())\n",
    "        \n",
    "        elif line.startswith(\"Sample\"):\n",
    "            result.append(lines[i + 1:i + 2])\n",
    "            \n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "parsed_output = {}\n",
    "vcfs = []\n",
    "\n",
    "for sample in sample_list:\n",
    "    parsed_output[sample] = {}\n",
    "    stellarpgx_files = [file for file in caller_outputs_path.glob(f\"{sample}/**/*.alleles\")]\n",
    "    stellarpgx_files_vcf = [file for file in caller_outputs_path.glob(f'{sample}/**/*.vcf.gz')]\n",
    "\n",
    "    for file in stellarpgx_files:\n",
    "        result = parse_output_stellarpgx(file)\n",
    "        parsed_output[sample][\"copy_number\"] = result[0]\n",
    "        parsed_output[sample][\"core_allele\"] = result[1]\n",
    "        parsed_output[sample][\"candidate_allele\"] = result[2]\n",
    "        parsed_output[sample][\"likely_background_allele\"] = result[-1]\n",
    "\n",
    "    for file in stellarpgx_files_vcf:       \n",
    "        vcf = pyvcf.VcfFrame.from_file(file.as_posix())        \n",
    "        vcfs.append(vcf)\n",
    "\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>samples</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WHB3386</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WHB3440</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WHB3461</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WHB3484</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WHB3488</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   samples\n",
       "0  WHB3386\n",
       "1  WHB3440\n",
       "2  WHB3461\n",
       "3  WHB3484\n",
       "4  WHB3488"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "novel_df = pd.DataFrame(parsed_output).T.reset_index(names='samples')\n",
    "novel_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Creating new columns for haplotypes\n",
    "novel_df['core_alleles_comparison'] = novel_df['core_allele'].apply(lambda x: re.findall(\"\\d{8}~\\w+>\\w+\", str(x))) # Core allele withour genotype\n",
    "novel_df['likely_background_allele'] = novel_df['likely_background_allele'].apply(lambda x: re.findall(\"\\*\\d+\", x))\n",
    "novel_df[['h1','h2']] = novel_df['likely_background_allele'].values.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0     [42126611~C>G, 42126877~G>A, 42127803~C>T, 421...\n",
       "1                                        [42130692~G>A]\n",
       "2     [42126611~C>G, 42127941~G>A, 42129779~A>G, 421...\n",
       "3            [42126611~C>G, 42127853~G>A, 42130692~G>A]\n",
       "4                                        [42130692~G>A]\n",
       "                            ...                        \n",
       "88                                       [42130692~G>A]\n",
       "89           [42126611~C>G, 42127853~G>A, 42130692~G>A]\n",
       "90           [42126611~C>G, 42127841~C>T, 42130692~G>A]\n",
       "91                                       [42130692~G>A]\n",
       "92                                       [42128341~C>T]\n",
       "Name: core_alleles_comparison, Length: 93, dtype: object"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "novel_df['core_alleles_comparison']"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Obtain the core variants of each haplotype from reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reference allele core variants\n",
    "ref_df = pd.read_csv('CYP2D6.NC_000022.11.haplotypes.tsv', sep = '\\t', skiprows= 1, usecols=['Haplotype Name','Variant Start','Reference Allele','Variant Allele'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Clean up the ref_dference allele core variants df\n",
    "ref_df['var'] = ref_df['Variant Start'].map(str) + \"~\" + ref_df['Reference Allele'].map(str) + \">\" + ref_df[\"Variant Allele\"].map(str)\n",
    "ref_df = ref_df.rename(columns = {'Haplotype Name':'haplotype'})\n",
    "\n",
    "# To keep only major allele core variants\n",
    "remove = ref_df.loc[ref_df['haplotype'].str.contains(\"\\.\")]\n",
    "ref_df.drop(remove.index, inplace =True)\n",
    "ref_df['haplotype'] = ref_df['haplotype'].str.replace(\"CYP2D6\", \"\")\n",
    "\n",
    "# Change to look up data frame to dictionary\n",
    "ref_dict = ref_df.groupby('haplotype').agg({'var': lambda x: ','.join(x)}).to_dict()['var']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Adding the core variants of haplotypes to the novel_df\n",
    "\n",
    "novel_df['h1_core_var'] = novel_df['h1'].apply(lambda x: ref_dict.get(x)).fillna(np.nan)\n",
    "novel_df['h2_core_var'] = novel_df['h2'].apply(lambda x: ref_dict.get(x)).fillna('[]')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>samples</th>\n",
       "      <th>copy_number</th>\n",
       "      <th>core_allele</th>\n",
       "      <th>candidate_allele</th>\n",
       "      <th>likely_background_allele</th>\n",
       "      <th>core_alleles_comparison</th>\n",
       "      <th>h1</th>\n",
       "      <th>h2</th>\n",
       "      <th>h1_core_var</th>\n",
       "      <th>h2_core_var</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WHB3386</td>\n",
       "      <td>2</td>\n",
       "      <td>[42126611~C&gt;G~1/1;42126877~G&gt;A~0/1;42127803~C&gt;...</td>\n",
       "      <td>[136.v1_4.v1]</td>\n",
       "      <td>[*136, *4]</td>\n",
       "      <td>[42126611~C&gt;G, 42126877~G&gt;A, 42127803~C&gt;T, 421...</td>\n",
       "      <td>*136</td>\n",
       "      <td>*4</td>\n",
       "      <td>42127941~G&gt;A,42126877~G&gt;A,42126611~C&gt;G</td>\n",
       "      <td>42128945~C&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WHB3440</td>\n",
       "      <td>3</td>\n",
       "      <td>[42130692~G&gt;A~0/1]</td>\n",
       "      <td>[10.v1_39.v1]</td>\n",
       "      <td>[*10, *39]</td>\n",
       "      <td>[42130692~G&gt;A]</td>\n",
       "      <td>*10</td>\n",
       "      <td>*39</td>\n",
       "      <td>42130692~G&gt;A,42126611~C&gt;G</td>\n",
       "      <td>42126611~C&gt;G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WHB3461</td>\n",
       "      <td>2</td>\n",
       "      <td>[42126611~C&gt;G~0/1;42127941~G&gt;A~0/1;42129779~A&gt;...</td>\n",
       "      <td>[34.v1_88.v1]</td>\n",
       "      <td>[*34, *88]</td>\n",
       "      <td>[42126611~C&gt;G, 42127941~G&gt;A, 42129779~A&gt;G, 421...</td>\n",
       "      <td>*34</td>\n",
       "      <td>*88</td>\n",
       "      <td>42127941~G&gt;A</td>\n",
       "      <td>42129779~A&gt;G,42126611~C&gt;G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WHB3484</td>\n",
       "      <td>3</td>\n",
       "      <td>[42126611~C&gt;G~0/1;42127853~G&gt;A~0/1;42130692~G&gt;...</td>\n",
       "      <td>[1.v1_10.v1]</td>\n",
       "      <td>[*1, *10]</td>\n",
       "      <td>[42126611~C&gt;G, 42127853~G&gt;A, 42130692~G&gt;A]</td>\n",
       "      <td>*1</td>\n",
       "      <td>*10</td>\n",
       "      <td>.~nan&gt;nan</td>\n",
       "      <td>42130692~G&gt;A,42126611~C&gt;G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WHB3488</td>\n",
       "      <td>3</td>\n",
       "      <td>[42130692~G&gt;A~0/1]</td>\n",
       "      <td>[10.v1_39.v1]</td>\n",
       "      <td>[*10, *39]</td>\n",
       "      <td>[42130692~G&gt;A]</td>\n",
       "      <td>*10</td>\n",
       "      <td>*39</td>\n",
       "      <td>42130692~G&gt;A,42126611~C&gt;G</td>\n",
       "      <td>42126611~C&gt;G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   samples copy_number                                        core_allele  \\\n",
       "0  WHB3386           2  [42126611~C>G~1/1;42126877~G>A~0/1;42127803~C>...   \n",
       "1  WHB3440           3                                 [42130692~G>A~0/1]   \n",
       "2  WHB3461           2  [42126611~C>G~0/1;42127941~G>A~0/1;42129779~A>...   \n",
       "3  WHB3484           3  [42126611~C>G~0/1;42127853~G>A~0/1;42130692~G>...   \n",
       "4  WHB3488           3                                 [42130692~G>A~0/1]   \n",
       "\n",
       "  candidate_allele likely_background_allele  \\\n",
       "0    [136.v1_4.v1]               [*136, *4]   \n",
       "1    [10.v1_39.v1]               [*10, *39]   \n",
       "2    [34.v1_88.v1]               [*34, *88]   \n",
       "3     [1.v1_10.v1]                [*1, *10]   \n",
       "4    [10.v1_39.v1]               [*10, *39]   \n",
       "\n",
       "                             core_alleles_comparison    h1   h2  \\\n",
       "0  [42126611~C>G, 42126877~G>A, 42127803~C>T, 421...  *136   *4   \n",
       "1                                     [42130692~G>A]   *10  *39   \n",
       "2  [42126611~C>G, 42127941~G>A, 42129779~A>G, 421...   *34  *88   \n",
       "3         [42126611~C>G, 42127853~G>A, 42130692~G>A]    *1  *10   \n",
       "4                                     [42130692~G>A]   *10  *39   \n",
       "\n",
       "                              h1_core_var                h2_core_var  \n",
       "0  42127941~G>A,42126877~G>A,42126611~C>G               42128945~C>T  \n",
       "1               42130692~G>A,42126611~C>G               42126611~C>G  \n",
       "2                            42127941~G>A  42129779~A>G,42126611~C>G  \n",
       "3                               .~nan>nan  42130692~G>A,42126611~C>G  \n",
       "4               42130692~G>A,42126611~C>G               42126611~C>G  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "novel_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "novel_df['h2_core_var'].isna().sum()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Add additional variant column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "add_var = []\n",
    "for i in range(novel_df.shape[0]):\n",
    "  try:\n",
    "    var = [x for x in novel_df['core_alleles_comparison'][i] if x not in novel_df['h2_core_var'][i] and x not in novel_df['h1_core_var'][i]]\n",
    "  except TypeError:\n",
    "    var = []\n",
    "  add_var.append(var)\n",
    "\n",
    "novel_df['add_var'] = add_var"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4.1 Merge vcfs and obtain AD information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>WHB3386</th>\n",
       "      <th>WHB3440</th>\n",
       "      <th>...</th>\n",
       "      <th>WHB5237</th>\n",
       "      <th>WHB5352</th>\n",
       "      <th>WHB5357</th>\n",
       "      <th>WHB5358</th>\n",
       "      <th>WHB5383</th>\n",
       "      <th>WHB5384</th>\n",
       "      <th>WHB5419</th>\n",
       "      <th>WHB5444</th>\n",
       "      <th>WHB5449</th>\n",
       "      <th>WHB5450</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>POS</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>42126310~C&gt;T</th>\n",
       "      <td>chr22</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:AD</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0/1:8,11</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42126347~T&gt;C</th>\n",
       "      <td>chr22</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:AD</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42126390~G&gt;A</th>\n",
       "      <td>chr22</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:AD</td>\n",
       "      <td>0/1:8,13</td>\n",
       "      <td>0/1:9,13</td>\n",
       "      <td>...</td>\n",
       "      <td>0/1:6,11</td>\n",
       "      <td>1/1:1,25</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0/1:9,9</td>\n",
       "      <td>0/1:7,17</td>\n",
       "      <td>0/1:6,8</td>\n",
       "      <td>0/1:6,10</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42126611~C&gt;G</th>\n",
       "      <td>chr22</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:AD</td>\n",
       "      <td>1/1:0,19</td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>1/1:0,22</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0/1:18,13</td>\n",
       "      <td></td>\n",
       "      <td>0/1:4,9</td>\n",
       "      <td>0/1:18,14</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42126741~A&gt;G</th>\n",
       "      <td>chr22</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>GT:AD</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 101 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              CHROM ID REF ALT QUAL FILTER INFO FORMAT   WHB3386   WHB3440  \\\n",
       "POS                                                                          \n",
       "42126310~C>T  chr22  .   C   T    .      .    .  GT:AD                       \n",
       "42126347~T>C  chr22  .   T   C    .      .    .  GT:AD                       \n",
       "42126390~G>A  chr22  .   G   A    .      .    .  GT:AD  0/1:8,13  0/1:9,13   \n",
       "42126611~C>G  chr22  .   C   G    .      .    .  GT:AD  1/1:0,19             \n",
       "42126741~A>G  chr22  .   A   G    .      .    .  GT:AD                       \n",
       "\n",
       "              ...   WHB5237   WHB5352 WHB5357 WHB5358    WHB5383  WHB5384  \\\n",
       "POS           ...                                                           \n",
       "42126310~C>T  ...                                       0/1:8,11            \n",
       "42126347~T>C  ...                                                           \n",
       "42126390~G>A  ...  0/1:6,11  1/1:1,25                             0/1:9,9   \n",
       "42126611~C>G  ...            1/1:0,22                  0/1:18,13            \n",
       "42126741~A>G  ...                                                           \n",
       "\n",
       "               WHB5419    WHB5444   WHB5449 WHB5450  \n",
       "POS                                                  \n",
       "42126310~C>T                                         \n",
       "42126347~T>C                                         \n",
       "42126390~G>A  0/1:7,17    0/1:6,8  0/1:6,10          \n",
       "42126611~C>G   0/1:4,9  0/1:18,14                    \n",
       "42126741~A>G                                         \n",
       "\n",
       "[5 rows x 101 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "novel_vcf = pyvcf.rescue_filtered_variants(vcfs, format = \"GT:AD\").df\n",
    "\n",
    "# Restructure the POS value to match the novel_df\n",
    "novel_vcf['POS'] = novel_vcf['POS'].map(str) + \"~\" + novel_vcf['REF'].map(str) + \">\" + novel_vcf[\"ALT\"].map(str)\n",
    "\n",
    "novel_vcf = novel_vcf.replace('./.:.', '').set_index('POS')\n",
    "\n",
    "novel_vcf.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4.2 Query AD information from novel_vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "ad_list = []\n",
    "for index, row in novel_df.iterrows():\n",
    "    sample = row['samples']\n",
    "    vars = row['add_var']\n",
    "    try :\n",
    "        value = novel_vcf[sample][vars].to_list()\n",
    "        ad_list.append(value)\n",
    "    except KeyError:\n",
    "        ad_list.append('LowABHet or LowQUAL')\n",
    "    \n",
    "\n",
    "\n",
    "\n",
    "novel_df['GT:AD'] = ad_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "novel_df.to_csv('novel_alleles.csv', index = False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cyp2d6",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
