{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Caller Consensus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data prep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "from collections import Counter\n",
    "from pathlib import Path\n",
    "from callers import Cyp2d6CallerOutput\n",
    "from diplotype import Diplotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "caller_outputs_path = Path(\"/home/torojr/SG10K-CYP2D6/_data/caller_outputs\")\n",
    "stellarpgx_outputs_path = caller_outputs_path / \"stellarpgx\"\n",
    "stellarpgx_sample_dirs = stellarpgx_outputs_path.glob(\"*\")\n",
    "sample_list = [sample.name for sample in stellarpgx_sample_dirs]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:root:More than one genotype found in Aldy output, only first solution will be processed\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[cyrius]Invalid diplotype string: \"*2_*39_*68\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:More than one genotype found in Aldy output, only first solution will be processed\n",
      "WARNING:root:[cyrius]Invalid diplotype string: \"*36_*4_*4\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:More than one genotype found in Aldy output, only first solution will be processed\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[cyrius]Invalid diplotype string: \"*1_*1_*83\", returning \"no_call\"\n",
      "WARNING:root:More than one genotype found in Aldy output, only first solution will be processed\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:More than one genotype found in Aldy output, only first solution will be processed\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[cyrius]Invalid diplotype string: \"*119_*2;*1_*41\", returning \"no_call\"\n",
      "WARNING:root:[cyrius]Invalid diplotype string: \"*10_*36_*36_*36_*36\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:More than one genotype found in Aldy output, only first solution will be processed\n",
      "WARNING:root:[cyrius]Invalid diplotype string: \"*10_*10_*68\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"*10/*41 or *39/*69\tDuplication present\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[cyrius]Invalid diplotype string: \"*119_*2;*1_*41\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"*10/*41 or *39/*69\tDuplication present\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[cyrius]Invalid diplotype string: \"*36_*41_*47\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[cyrius]Invalid diplotype string: \"*10_*4_*4.013;*36_*4_*4\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[cyrius]Invalid diplotype string: \"*1_*36_*49\", returning \"no_call\"\n",
      "WARNING:root:[cyrius]Invalid diplotype string: \"*1_*1_*68\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[cyrius]Invalid diplotype string: \"*10_*36_*36_*36_*36\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:More than one genotype found in Aldy output, only first solution will be processed\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"*5/*other\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[cyrius]Invalid diplotype string: \"*36_*4_*4\", returning \"no_call\"\n",
      "WARNING:root:More than one genotype found in Aldy output, only first solution will be processed\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:More than one genotype found in Aldy output, only first solution will be processed\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"*5/*other\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[cyrius]Invalid diplotype string: \"*1_*10_*68\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:More than one genotype found in Aldy output, only first solution will be processed\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"*5/*other\", returning \"no_call\"\n",
      "WARNING:root:More than one genotype found in Aldy output, only first solution will be processed\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[cyrius]Invalid diplotype string: \"*1_*2_*68\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"*10/*41 or *39/*69\tDuplication present\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[cyrius]Invalid diplotype string: \"*119_*2;*1_*41\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[cyrius]Invalid diplotype string: \"*10_*10_*68\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:More than one genotype found in Aldy output, only first solution will be processed\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[cyrius]Invalid diplotype string: \"*119_*2;*1_*41\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[cyrius]Invalid diplotype string: \"*1_*1_*83\", returning \"no_call\"\n",
      "WARNING:root:More than one genotype found in Aldy output, only first solution will be processed\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"*5/*other\", returning \"no_call\"\n",
      "WARNING:root:More than one genotype found in Aldy output, only first solution will be processed\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"*5/*other\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n",
      "WARNING:root:[stellarpgx]Invalid diplotype string: \"Possible novel allele or suballele present: interpret with caution; experimental validation and expert review through PharmVar is recommended\", returning \"no_call\"\n"
     ]
    }
   ],
   "source": [
    "diplotypes = {}\n",
    "\n",
    "for sample in sample_list:\n",
    "    diplotypes[sample] = {}\n",
    "    files = {\n",
    "        \"cyrius\": [file for file in caller_outputs_path.glob(f\"sg10k-dragen/{sample}/cyrius/*.json\")][0],\n",
    "        \"aldy\": [file for file in caller_outputs_path.glob(f\"sg10k-dragen/{sample}/aldy/*.aldy\")][0],\n",
    "        \"stellarpgx\": [file for file in caller_outputs_path.glob(f\"stellarpgx/{sample}/**/*.alleles\")][0]\n",
    "    }\n",
    "    \n",
    "    for caller, file in files.items():\n",
    "        caller_output = eval(f\"Cyp2d6CallerOutput.{caller}(\\\"{file}\\\")\")\n",
    "        diplotypes[sample][f\"raw_{caller}\"] = caller_output.diplotype.raw\n",
    "        diplotypes[sample][caller] = caller_output.diplotype.parsed"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculate consensus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def simple_consensus(cyrius, aldy, stellarpgx):\n",
    "    consensus = None\n",
    "    if cyrius == aldy == stellarpgx:\n",
    "        consensus = cyrius\n",
    "    elif cyrius == aldy:\n",
    "        consensus = cyrius\n",
    "    elif cyrius == stellarpgx:\n",
    "        consensus =  cyrius\n",
    "    elif aldy == stellarpgx:\n",
    "        consensus = aldy\n",
    "    return consensus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "diplotypes_df = pd.DataFrame.from_dict(diplotypes, orient=\"index\")\n",
    "diplotypes_df[\"all_match\"] = (diplotypes_df[\"cyrius\"] == diplotypes_df[\"aldy\"]) & (diplotypes_df[\"cyrius\"] == diplotypes_df[\"stellarpgx\"])\n",
    "diplotypes_df[\"diplotype\"] = diplotypes_df.apply(lambda sample: simple_consensus(sample[\"cyrius\"], sample[\"aldy\"], sample[\"stellarpgx\"]), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#diplotypes_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = pd.read_csv(\"/home/torojr/SG10K-CYP2D6/_data/metadata/SG10K-DRAGEN.sample_list_full.20210719.csv\")\n",
    "metadata.set_index(\"npm_research_id\", inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "diplotypes_with_metadata = pd.merge(diplotypes_df, metadata, how=\"left\", left_index=True, right_index=True,)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#diplotypes_with_metadata.sample(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Consensus aggregate data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "consensus_df = diplotypes_with_metadata.loc[(diplotypes_with_metadata[\"diplotype\"].notna()) & (diplotypes_with_metadata[\"diplotype\"] != \"no_call\")]\n",
    "#consensus_df.sample(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "genetic_ancestry\n",
       "C    1154\n",
       "I     166\n",
       "M     203\n",
       "O       2\n",
       "Name: diplotype, dtype: int64"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ethnicity_counts = consensus_df.groupby(\"genetic_ancestry\")[\"diplotype\"].count()\n",
    "ethnicity_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1525"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ethnicity_counts.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>diplotype</th>\n",
       "      <th>genetic_ancestry</th>\n",
       "      <th>diplotype_count</th>\n",
       "      <th>diplotype_frequency</th>\n",
       "      <th>haplotypes</th>\n",
       "      <th>star_alleles</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>*41/*41</td>\n",
       "      <td>M</td>\n",
       "      <td>1</td>\n",
       "      <td>0.004926</td>\n",
       "      <td>[*41, *41]</td>\n",
       "      <td>[*41, *41]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>*10+*36/*14</td>\n",
       "      <td>C</td>\n",
       "      <td>9</td>\n",
       "      <td>0.007799</td>\n",
       "      <td>[*10+*36, *14]</td>\n",
       "      <td>[*10, *36, *14]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>*5/*71</td>\n",
       "      <td>C</td>\n",
       "      <td>3</td>\n",
       "      <td>0.002600</td>\n",
       "      <td>[*5, *71]</td>\n",
       "      <td>[*5, *71]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>*10/*41</td>\n",
       "      <td>I</td>\n",
       "      <td>1</td>\n",
       "      <td>0.006024</td>\n",
       "      <td>[*10, *41]</td>\n",
       "      <td>[*10, *41]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>*4/*41</td>\n",
       "      <td>I</td>\n",
       "      <td>3</td>\n",
       "      <td>0.018072</td>\n",
       "      <td>[*4, *41]</td>\n",
       "      <td>[*4, *41]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       diplotype genetic_ancestry  diplotype_count  diplotype_frequency  \\\n",
       "144      *41/*41                M                1             0.004926   \n",
       "50   *10+*36/*14                C                9             0.007799   \n",
       "161       *5/*71                C                3             0.002600   \n",
       "79       *10/*41                I                1             0.006024   \n",
       "138       *4/*41                I                3             0.018072   \n",
       "\n",
       "         haplotypes     star_alleles  \n",
       "144      [*41, *41]       [*41, *41]  \n",
       "50   [*10+*36, *14]  [*10, *36, *14]  \n",
       "161       [*5, *71]        [*5, *71]  \n",
       "79       [*10, *41]       [*10, *41]  \n",
       "138       [*4, *41]        [*4, *41]  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aggregate_df = consensus_df.groupby([\"diplotype\", \"genetic_ancestry\"])[\"diplotype\"].count().to_frame().rename(columns={\"diplotype\": \"diplotype_count\"}).reset_index()\n",
    "aggregate_df[\"diplotype_frequency\"] = aggregate_df.apply(lambda x: x[\"diplotype_count\"] / ethnicity_counts.loc[x[\"genetic_ancestry\"]], axis=1)\n",
    "aggregate_df[\"haplotypes\"] = aggregate_df[\"diplotype\"].str.split(\"/\")\n",
    "aggregate_df[\"star_alleles\"] = aggregate_df[\"diplotype\"].apply(Diplotype.from_string).apply(lambda x: list(pd.Series([list(h.star_alleles.elements()) for h in x.haplotypes]).explode()))\n",
    "aggregate_df.sample(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "haplotype_counts_by_ethnicity = {}\n",
    "star_allele_counts_by_ethnicity = {}\n",
    "for ethnicity, data in aggregate_df.groupby(\"genetic_ancestry\"):\n",
    "    haplotype_counts_by_ethnicity[ethnicity] = Counter(data[\"haplotypes\"].explode())\n",
    "    star_allele_counts_by_ethnicity[ethnicity] = Counter(data[\"star_alleles\"].explode())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>diplotype</th>\n",
       "      <th>genetic_ancestry</th>\n",
       "      <th>diplotype_count</th>\n",
       "      <th>diplotype_frequency</th>\n",
       "      <th>haplotypes</th>\n",
       "      <th>star_alleles</th>\n",
       "      <th>haplotype_count</th>\n",
       "      <th>haplotype_frequency</th>\n",
       "      <th>star_allele_count</th>\n",
       "      <th>star_allele_frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>*2x2/*41</td>\n",
       "      <td>M</td>\n",
       "      <td>1</td>\n",
       "      <td>0.004926</td>\n",
       "      <td>[*2x2, *41]</td>\n",
       "      <td>[*2, *2, *41]</td>\n",
       "      <td>[2, 8]</td>\n",
       "      <td>[0.02702702702702703, 0.10810810810810811]</td>\n",
       "      <td>[11, 11, 8]</td>\n",
       "      <td>[0.11956521739130435, 0.11956521739130435, 0.0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>*2x2/*10</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000867</td>\n",
       "      <td>[*2x2, *10]</td>\n",
       "      <td>[*2, *2, *10]</td>\n",
       "      <td>[4, 15]</td>\n",
       "      <td>[0.02631578947368421, 0.09868421052631579]</td>\n",
       "      <td>[22, 22, 45]</td>\n",
       "      <td>[0.11224489795918367, 0.11224489795918367, 0.2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>*1x2/*41</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000867</td>\n",
       "      <td>[*1x2, *41]</td>\n",
       "      <td>[*1, *1, *41]</td>\n",
       "      <td>[4, 13]</td>\n",
       "      <td>[0.02631578947368421, 0.08552631578947369]</td>\n",
       "      <td>[26, 26, 13]</td>\n",
       "      <td>[0.1326530612244898, 0.1326530612244898, 0.066...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>*1/*10+*36x2</td>\n",
       "      <td>M</td>\n",
       "      <td>1</td>\n",
       "      <td>0.004926</td>\n",
       "      <td>[*1, *10+*36x2]</td>\n",
       "      <td>[*1, *10, *36, *36]</td>\n",
       "      <td>[12, 2]</td>\n",
       "      <td>[0.16216216216216217, 0.02702702702702703]</td>\n",
       "      <td>[14, 25, 15, 15]</td>\n",
       "      <td>[0.15217391304347827, 0.2717391304347826, 0.16...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>*14/*36x2</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000867</td>\n",
       "      <td>[*14, *36x2]</td>\n",
       "      <td>[*14, *36, *36]</td>\n",
       "      <td>[7, 1]</td>\n",
       "      <td>[0.046052631578947366, 0.006578947368421052]</td>\n",
       "      <td>[7, 36, 36]</td>\n",
       "      <td>[0.03571428571428571, 0.1836734693877551, 0.18...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        diplotype genetic_ancestry  diplotype_count  diplotype_frequency  \\\n",
       "123      *2x2/*41                M                1             0.004926   \n",
       "121      *2x2/*10                C                1             0.000867   \n",
       "94       *1x2/*41                C                1             0.000867   \n",
       "10   *1/*10+*36x2                M                1             0.004926   \n",
       "88      *14/*36x2                C                1             0.000867   \n",
       "\n",
       "          haplotypes         star_alleles haplotype_count  \\\n",
       "123      [*2x2, *41]        [*2, *2, *41]          [2, 8]   \n",
       "121      [*2x2, *10]        [*2, *2, *10]         [4, 15]   \n",
       "94       [*1x2, *41]        [*1, *1, *41]         [4, 13]   \n",
       "10   [*1, *10+*36x2]  [*1, *10, *36, *36]         [12, 2]   \n",
       "88      [*14, *36x2]      [*14, *36, *36]          [7, 1]   \n",
       "\n",
       "                              haplotype_frequency star_allele_count  \\\n",
       "123    [0.02702702702702703, 0.10810810810810811]       [11, 11, 8]   \n",
       "121    [0.02631578947368421, 0.09868421052631579]      [22, 22, 45]   \n",
       "94     [0.02631578947368421, 0.08552631578947369]      [26, 26, 13]   \n",
       "10     [0.16216216216216217, 0.02702702702702703]  [14, 25, 15, 15]   \n",
       "88   [0.046052631578947366, 0.006578947368421052]       [7, 36, 36]   \n",
       "\n",
       "                                 star_allele_frequency  \n",
       "123  [0.11956521739130435, 0.11956521739130435, 0.0...  \n",
       "121  [0.11224489795918367, 0.11224489795918367, 0.2...  \n",
       "94   [0.1326530612244898, 0.1326530612244898, 0.066...  \n",
       "10   [0.15217391304347827, 0.2717391304347826, 0.16...  \n",
       "88   [0.03571428571428571, 0.1836734693877551, 0.18...  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aggregate_df[\"haplotype_count\"] = aggregate_df.apply(lambda x: [haplotype_counts_by_ethnicity[x[\"genetic_ancestry\"]].get(haplotype) for haplotype in x[\"haplotypes\"]], axis=1)\n",
    "aggregate_df[\"haplotype_frequency\"] = aggregate_df.apply(lambda x: [hc / haplotype_counts_by_ethnicity[x[\"genetic_ancestry\"]].total() for hc in x[\"haplotype_count\"]], axis=1)\n",
    "aggregate_df[\"star_allele_count\"] = aggregate_df.apply(lambda x: [star_allele_counts_by_ethnicity[x[\"genetic_ancestry\"]].get(star_allele) for star_allele in x[\"star_alleles\"]], axis=1)\n",
    "aggregate_df[\"star_allele_frequency\"] = aggregate_df.apply(lambda x: [sac / star_allele_counts_by_ethnicity[x[\"genetic_ancestry\"]].total() for sac in x[\"star_allele_count\"]], axis=1)\n",
    "aggregate_df[aggregate_df[\"diplotype\"].str.contains(\"x\")].sample(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export\n",
    "columns = [\"genetic_ancestry\", \"diplotype\", \"diplotype_count\", \"diplotype_frequency\", \"haplotypes\", \"haplotype_count\", \"haplotype_frequency\", \"star_alleles\", \"star_allele_count\", \"star_allele_frequency\"]\n",
    "aggregate_df = aggregate_df.loc[:, columns]\n",
    "aggregate_df.sort_values([\"diplotype_count\"], ascending=False).reset_index(drop=True).to_csv(\"/home/torojr/SG10K-CYP2D6/_data/sg10k_cyp2d6_diplotype_consensus_v1.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['*1/*1', '*1/*10', '*1/*10+*36', '*1/*10+*36x2', '*1/*113',\n",
       "       '*1/*14', '*1/*17', '*1/*1x2', '*1/*2', '*1/*21', '*1/*2x2',\n",
       "       '*1/*33', '*1/*35', '*1/*36', '*1/*4', '*1/*4+*68', '*1/*41',\n",
       "       '*1/*43', '*1/*49', '*1/*4x2', '*1/*5', '*1/*71', '*1/*94',\n",
       "       '*10+*36/*10+*36', '*10+*36/*10+*36x2', '*10+*36/*133',\n",
       "       '*10+*36/*14', '*10+*36/*21', '*10+*36/*35', '*10+*36/*36',\n",
       "       '*10+*36/*41', '*10+*36/*43', '*10+*36/*49', '*10+*36/*49x2',\n",
       "       '*10+*36/*65', '*10+*36/*71', '*10+*36x2/*35', '*10+*36x2/*41',\n",
       "       '*10/*10', '*10/*10+*36', '*10/*111', '*10/*14', '*10/*35',\n",
       "       '*10/*36', '*10/*41', '*10/*49', '*10/*52', '*10/*71', '*10/*75',\n",
       "       '*10x2/*10+*36', '*10x2/*94', '*14/*36x2', '*14/*41',\n",
       "       '*1x2/*10+*36', '*1x2/*2', '*1x2/*4', '*1x2/*41', '*2/*10',\n",
       "       '*2/*10+*36', '*2/*10+*36x2', '*2/*14', '*2/*2', '*2/*2x2',\n",
       "       '*2/*3', '*2/*35', '*2/*4', '*2/*4+*68', '*2/*41', '*2/*43',\n",
       "       '*2/*49', '*2/*5', '*2/*71', '*2x2/*10', '*2x2/*10+*36',\n",
       "       '*2x2/*41', '*2x2/*71', '*35/*41', '*35/*86', '*39/*41', '*39/*69',\n",
       "       '*4+*68/*41', '*4/*10', '*4/*10+*36', '*4/*10+*36x2', '*4/*112',\n",
       "       '*4/*41', '*4/*5', '*4/*82', '*4/*86', '*41/*41', '*41/*43',\n",
       "       '*41/*49', '*41/*94', '*5/*10', '*5/*111', '*5/*14', '*5/*15',\n",
       "       '*5/*21', '*5/*41', '*5/*49', '*5/*5', '*5/*7', '*5/*71', '*5/*86',\n",
       "       '*5/*9'], dtype=object)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aggregate_df[\"diplotype\"].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Samples without consensus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>raw_cyrius</th>\n",
       "      <th>cyrius</th>\n",
       "      <th>raw_aldy</th>\n",
       "      <th>aldy</th>\n",
       "      <th>raw_stellarpgx</th>\n",
       "      <th>stellarpgx</th>\n",
       "      <th>all_match</th>\n",
       "      <th>diplotype</th>\n",
       "      <th>multiplex_pool_id</th>\n",
       "      <th>supplier_id</th>\n",
       "      <th>...</th>\n",
       "      <th>supplied_gender</th>\n",
       "      <th>genetic_sex</th>\n",
       "      <th>self_reported_ethnicity</th>\n",
       "      <th>genetic_ancestry</th>\n",
       "      <th>extraction_kit</th>\n",
       "      <th>library_prep_kit</th>\n",
       "      <th>sequencing_depth</th>\n",
       "      <th>estimate_of_sequence_coverage</th>\n",
       "      <th>qc_pass_in_r5_3</th>\n",
       "      <th>industry_consent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>WHB4651</th>\n",
       "      <td>no_call</td>\n",
       "      <td>no_call</td>\n",
       "      <td>*10+*10+*10+*36.ALDY+*10/*36.ALDY+*10</td>\n",
       "      <td>*10x4+*36/*10+*36</td>\n",
       "      <td>*10/*10x7</td>\n",
       "      <td>*10/*10x7</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>MUX9238</td>\n",
       "      <td>010-20832</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>male</td>\n",
       "      <td>Malay</td>\n",
       "      <td>M</td>\n",
       "      <td>QIAsymphony DSP DNA Midi Kit</td>\n",
       "      <td>NEBNext UltraII DNA library Prep Kit for Illumina</td>\n",
       "      <td>30x</td>\n",
       "      <td>27.614180</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WHB3671</th>\n",
       "      <td>*36+*10/*36+*10</td>\n",
       "      <td>*10+*36/*10+*36</td>\n",
       "      <td>*10/*36.ALDY+*10</td>\n",
       "      <td>*10/*10+*36</td>\n",
       "      <td>*10/*10x3</td>\n",
       "      <td>*10/*10x3</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>MUX8198</td>\n",
       "      <td>0217-0012</td>\n",
       "      <td>...</td>\n",
       "      <td>F</td>\n",
       "      <td>female</td>\n",
       "      <td>Chinese</td>\n",
       "      <td>C</td>\n",
       "      <td>Chemagic DNA Blood Kit (Perkin Elmer, MA)</td>\n",
       "      <td>NEBNext UltraII DNA library Prep Kit for Illumina</td>\n",
       "      <td>30x</td>\n",
       "      <td>28.542650</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WHB4552</th>\n",
       "      <td>no_call</td>\n",
       "      <td>no_call</td>\n",
       "      <td>*5/*41</td>\n",
       "      <td>*5/*41</td>\n",
       "      <td>*41/*41</td>\n",
       "      <td>*41/*41</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>MUX9208</td>\n",
       "      <td>020-04180</td>\n",
       "      <td>...</td>\n",
       "      <td>F</td>\n",
       "      <td>female</td>\n",
       "      <td>Indian</td>\n",
       "      <td>I</td>\n",
       "      <td>QIAsymphony DSP DNA Midi Kit</td>\n",
       "      <td>NEBNext UltraII DNA library Prep Kit for Illumina</td>\n",
       "      <td>30x</td>\n",
       "      <td>29.041328</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WHB4740</th>\n",
       "      <td>*10/*2</td>\n",
       "      <td>*2/*10</td>\n",
       "      <td>*2/*36.ALDY+*10</td>\n",
       "      <td>*2/*10+*36</td>\n",
       "      <td>*2/*36x2+*10</td>\n",
       "      <td>*2/*10+*36x2</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>MUX9276</td>\n",
       "      <td>010-21410</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>male</td>\n",
       "      <td>Malay</td>\n",
       "      <td>M</td>\n",
       "      <td>QIAsymphony DSP DNA Midi Kit</td>\n",
       "      <td>NEBNext UltraII DNA library Prep Kit for Illumina</td>\n",
       "      <td>30x</td>\n",
       "      <td>29.284166</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WHB5126</th>\n",
       "      <td>*14/*36+*10</td>\n",
       "      <td>*10+*36/*14</td>\n",
       "      <td>*2+*36.ALDY+*10/*14+*68</td>\n",
       "      <td>*2+*10+*36/*14+*68</td>\n",
       "      <td>*2/*36+*10</td>\n",
       "      <td>*2/*10+*36</td>\n",
       "      <td>False</td>\n",
       "      <td>None</td>\n",
       "      <td>MUX8358</td>\n",
       "      <td>1217-0005</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>male</td>\n",
       "      <td>Chinese</td>\n",
       "      <td>C</td>\n",
       "      <td>Chemagic DNA Blood Kit (Perkin Elmer, MA)</td>\n",
       "      <td>NEBNext UltraII DNA library Prep Kit for Illumina</td>\n",
       "      <td>30x</td>\n",
       "      <td>31.346788</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              raw_cyrius           cyrius  \\\n",
       "WHB4651          no_call          no_call   \n",
       "WHB3671  *36+*10/*36+*10  *10+*36/*10+*36   \n",
       "WHB4552          no_call          no_call   \n",
       "WHB4740           *10/*2           *2/*10   \n",
       "WHB5126      *14/*36+*10      *10+*36/*14   \n",
       "\n",
       "                                      raw_aldy                aldy  \\\n",
       "WHB4651  *10+*10+*10+*36.ALDY+*10/*36.ALDY+*10   *10x4+*36/*10+*36   \n",
       "WHB3671                       *10/*36.ALDY+*10         *10/*10+*36   \n",
       "WHB4552                                 *5/*41              *5/*41   \n",
       "WHB4740                        *2/*36.ALDY+*10          *2/*10+*36   \n",
       "WHB5126                *2+*36.ALDY+*10/*14+*68  *2+*10+*36/*14+*68   \n",
       "\n",
       "        raw_stellarpgx    stellarpgx  all_match diplotype multiplex_pool_id  \\\n",
       "WHB4651      *10/*10x7     *10/*10x7      False      None           MUX9238   \n",
       "WHB3671      *10/*10x3     *10/*10x3      False      None           MUX8198   \n",
       "WHB4552        *41/*41       *41/*41      False      None           MUX9208   \n",
       "WHB4740   *2/*36x2+*10  *2/*10+*36x2      False      None           MUX9276   \n",
       "WHB5126     *2/*36+*10    *2/*10+*36      False      None           MUX8358   \n",
       "\n",
       "        supplier_id  ... supplied_gender genetic_sex self_reported_ethnicity  \\\n",
       "WHB4651   010-20832  ...               M        male                   Malay   \n",
       "WHB3671   0217-0012  ...               F      female                 Chinese   \n",
       "WHB4552   020-04180  ...               F      female                  Indian   \n",
       "WHB4740   010-21410  ...               M        male                   Malay   \n",
       "WHB5126   1217-0005  ...               M        male                 Chinese   \n",
       "\n",
       "        genetic_ancestry                             extraction_kit  \\\n",
       "WHB4651                M               QIAsymphony DSP DNA Midi Kit   \n",
       "WHB3671                C  Chemagic DNA Blood Kit (Perkin Elmer, MA)   \n",
       "WHB4552                I               QIAsymphony DSP DNA Midi Kit   \n",
       "WHB4740                M               QIAsymphony DSP DNA Midi Kit   \n",
       "WHB5126                C  Chemagic DNA Blood Kit (Perkin Elmer, MA)   \n",
       "\n",
       "                                          library_prep_kit sequencing_depth  \\\n",
       "WHB4651  NEBNext UltraII DNA library Prep Kit for Illumina              30x   \n",
       "WHB3671  NEBNext UltraII DNA library Prep Kit for Illumina              30x   \n",
       "WHB4552  NEBNext UltraII DNA library Prep Kit for Illumina              30x   \n",
       "WHB4740  NEBNext UltraII DNA library Prep Kit for Illumina              30x   \n",
       "WHB5126  NEBNext UltraII DNA library Prep Kit for Illumina              30x   \n",
       "\n",
       "        estimate_of_sequence_coverage  qc_pass_in_r5_3  industry_consent  \n",
       "WHB4651                     27.614180             True              True  \n",
       "WHB3671                     28.542650             True             False  \n",
       "WHB4552                     29.041328             True              True  \n",
       "WHB4740                     29.284166             True              True  \n",
       "WHB5126                     31.346788             True             False  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "no_consensus_df = diplotypes_with_metadata[diplotypes_with_metadata[\"diplotype\"].isna()]\n",
    "no_consensus_df.sample(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "C    200\n",
       "M     64\n",
       "I     27\n",
       "Name: genetic_ancestry, dtype: int64"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "no_consensus_df[\"genetic_ancestry\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False    147\n",
       "True     144\n",
       "Name: cyrius, dtype: int64"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(no_consensus_df[\"cyrius\"] == \"no_call\").value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['aldy,stellarpgx',\n",
       " '*2/*27,*2/*2',\n",
       " '*10/*10+*36,*10+*36/*10+*36',\n",
       " '*1/*2,*1/*2x2',\n",
       " '*1/*2,*1/*2x2',\n",
       " '*10x7+*36/*10+*36,*10/*10x6',\n",
       " '*10+*36/*10+*36,*10/*10+*36',\n",
       " '*10x2+*36/*10+*36,*10/*10x3',\n",
       " '*1/*1,*1/*1x2',\n",
       " '*1/*41,*1/*41x2',\n",
       " '*2/*10+*36,*2/*10x2+*36',\n",
       " '*1/*36x2,*1/*10+*36',\n",
       " '*1/*2x2,*1/*2x3',\n",
       " '*2+*68/*39,*2/*10+*36',\n",
       " '*1/*5,*1/*1',\n",
       " '*4/*41,*4x2/*41',\n",
       " '*10x4+*36/*10+*36,*10/*10x5',\n",
       " '*4/*10+*36,*4x2/*10',\n",
       " '*1/*10+*36,*1/*10+*36x2',\n",
       " '*10+*36/*10+*36,*10+*36/*10+*36x2',\n",
       " '*4/*10+*36,*4x3/*10',\n",
       " '*1/*10,*1/*10x2',\n",
       " '*1+*36/*10+*36,*1/*10x2+*36x2',\n",
       " '*1/*10+*36,*1/*10+*36x2',\n",
       " '*1x2/*61,*1/*1x2',\n",
       " '*1/*10,*1/*10x2',\n",
       " '*10/*10+*36,*10/*10x2',\n",
       " '*10+*36/*36,*10+*36/*10+*36',\n",
       " '*1/*10,*1/*10x2',\n",
       " '*1/*10,*1/*10x2',\n",
       " '*10+*36/*10+*36,*10+*36/*10x2+*36',\n",
       " '*1/*41,*1/*41x2',\n",
       " '*1/*61,*1/*143',\n",
       " '*1x2+*10+*36/*10+*36,*1/*10+*36x2',\n",
       " '*10/*10,*10/*10x2',\n",
       " '*1/*4,*1/*4x2',\n",
       " '*5/*119,*34x2/*119',\n",
       " '*4/*10,*4x2/*10',\n",
       " '*1/*2,*1/*2x2',\n",
       " '*10x2/*68,*10/*10+*36',\n",
       " '*2/*10+*36,*2/*10',\n",
       " '*36x2/*65x2,*10/*65x2',\n",
       " '*10/*10+*36,*10+*36/*10+*36',\n",
       " '*10/*41,*10/*41x2',\n",
       " '*2/*119,*1/*41x2',\n",
       " '*10+*36/*10+*36,*10+*36/*10x2+*36',\n",
       " '*2/*10+*36,*2/*10x2+*36',\n",
       " '*10x2/*68+*83,*10+*36/*39',\n",
       " '*36+*47/*41,*41/*47x2',\n",
       " '*1/*41,*1/*41x2',\n",
       " '*4/*5,*4/*4',\n",
       " '*1/*10,*1/*10x2',\n",
       " '*2/*10,*2x2/*10',\n",
       " '*2/*10,*2x2/*10',\n",
       " '*10+*36/*10+*36,*10/*10x5',\n",
       " '*1+*49/*36,*1/*49x2',\n",
       " '*1/*10+*36,*1/*10x2+*36',\n",
       " '*5/*83,*1/*1x2',\n",
       " '*10/*10+*36,*10/*10x2+*36',\n",
       " '*10x2+*36/*10+*36,*10+*36/*10+*36',\n",
       " '*1/*36x2,*1/*10+*36',\n",
       " '*41/*41,*41/*41x2',\n",
       " '*10+*63/*114,*10/*14x2',\n",
       " '*1/*2,*1/*2x2',\n",
       " '*1/*10+*36,*1/*10x2+*36',\n",
       " '*10+*36/*10+*36,*10+*36/*10+*36x2',\n",
       " '*4/*10+*36,*4x2/*10',\n",
       " '*10/*10,*10/*10x2',\n",
       " '*10/*10+*36,*10/*10x3+*36',\n",
       " '*1/*10+*36,*1/*10+*36x2',\n",
       " '*1/*10,*1/*10x2',\n",
       " '*10/*36,*10/*10+*36',\n",
       " '*10/*10+*36,*10/*10x2+*36',\n",
       " '*4/*10+*36,*4/*10x2',\n",
       " '*10+*36/*10+*36,*10/*10x3',\n",
       " '*10+*36/*36,*10/*10+*36',\n",
       " '*4/*86,*4+*68/*86',\n",
       " '*10+*36/*10+*36,*10/*36x2',\n",
       " '*10+*22/*10+*36,*10+*36/*142x2',\n",
       " '*1+*68/*10,*1/*10+*36',\n",
       " '*1/*10+*36,*1/*10+*36x2',\n",
       " '*2/*2,*2/*2x2',\n",
       " '*1+*61/*10+*36,*1/*10+*36x3',\n",
       " '*1/*5,*1/*1',\n",
       " '*1/*10+*36,*1/*10x2+*36x2',\n",
       " '*1/*10,*1/*10x2',\n",
       " '*2/*10+*36,*2/*10',\n",
       " '*2/*10+*36,*2/*10x2+*36',\n",
       " '*10+*36/*10+*36,*10/*10+*36',\n",
       " '*10+*36/*10+*36,*10/*10+*36',\n",
       " '*1/*1,*1/*1x2',\n",
       " '*2/*10+*36,*2/*10x2+*36',\n",
       " '*4/*43,*4x2/*43',\n",
       " '*10x4+*36/*10+*36,*10/*10x7',\n",
       " '*10+*36/*10+*36,*10/*10x5',\n",
       " '*1x2+*10+*36/*10x2+*36,*1/*10+*36x3',\n",
       " '*1/*5,*1/*2x2',\n",
       " '*1/*41,*1/*41x2',\n",
       " '*10/*10+*36,*10+*36/*10+*36',\n",
       " '*63/*80,*2/*41',\n",
       " '*10/*10+*36,*10/*10',\n",
       " '*1x2+*10+*36/*10x2+*36,*1/*10+*36x3',\n",
       " '*10/*10+*36,*10+*36/*10+*36',\n",
       " '*2/*10,*2x2/*10',\n",
       " '*1/*41,*1/*41x2',\n",
       " '*35/*41,*35/*41x2',\n",
       " '*10/*10,*10/*10x2',\n",
       " '*4+*68/*36,*4x2/*10+*36',\n",
       " '*1/*10,*1/*10x2',\n",
       " '*10/*36,*10/*10+*36',\n",
       " '*10/*10+*36,*10+*36/*10+*36',\n",
       " '*1/*1,*1/*1x2',\n",
       " '*10/*36,*10/*36x2',\n",
       " '*10x3/*10+*36+*83,*10x2+*36x3/*39',\n",
       " '*10+*36/*10+*36,*10/*10+*36',\n",
       " '*10+*36/*10+*36,*10/*36x2',\n",
       " '*5/*10,*10/*36',\n",
       " '*1/*10,*1/*10x2',\n",
       " '*10+*36/*79,*10/*69',\n",
       " '*10/*107,*10/*107x2',\n",
       " '*2/*43,*2x2/*43',\n",
       " '*1/*41,*34/*119x2',\n",
       " '*2/*36x2,*2/*10+*36',\n",
       " '*10+*36/*10+*36+*41,*10x2+*36x2/*41',\n",
       " '*1/*49,*1/*49x2',\n",
       " '*1/*10+*36,*10+*36/*143',\n",
       " '*10+*36/*10+*36,*10/*10+*36',\n",
       " '*1/*1,*1/*1x2',\n",
       " '*1x2/*61,*1/*1',\n",
       " '*5/*41,*41/*41',\n",
       " '*10+*36/*10+*36,*10/*10x5',\n",
       " '*1/*1,*1/*1x2',\n",
       " '*1/*10+*36,*1/*10x2+*36',\n",
       " '*1+*36/*10+*36,*1/*36x2',\n",
       " '*10+*36/*49,*10+*36/*49x2',\n",
       " '*5/*5,*2/*2x2',\n",
       " '*2/*10,*2x2/*10',\n",
       " '*10/*49,*10/*49x2',\n",
       " '*1+*36/*10+*36,*1/*10+*36',\n",
       " '*2/*4,*2/*4x2',\n",
       " '*10+*36/*10+*36,*10/*36x2',\n",
       " '*1/*10+*36,*1/*10+*36x2',\n",
       " '*10/*10+*36,*10/*10x2+*36',\n",
       " '*1x2+*10+*36/*10x4+*36,*1/*10+*36x4',\n",
       " '*10x3/*10+*36+*83,*10+*36x3/*39']"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Cyrius \"no_call\"\n",
    "no_consensus_df[no_consensus_df[\"cyrius\"] == \"no_call\"][[\"aldy\", \"stellarpgx\"]].to_csv(index=False).split()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['cyrius,aldy,stellarpgx',\n",
       " '*10+*36/*10+*36,*10x2+*36/*10+*36,*10/*10+*36',\n",
       " '*10+*36/*36x2,*10+*36/*10+*36,*10/*10x3',\n",
       " '*10/*10+*36,*10+*36/*36,*10/*36x2',\n",
       " '*10+*36/*36x2,*10+*36/*10+*36,*10/*36x2',\n",
       " '*10+*36/*36x2,*10+*36/*10+*36,*10/*10x3',\n",
       " '*2/*10,*2/*10+*36,*2/*10x2+*36',\n",
       " '*1/*10+*36x2,*1+*36/*10+*36,*1/*10+*36',\n",
       " '*1/*13,*1/*2,*1/*2x2',\n",
       " '*10+*36/*10+*36x2,*10x3+*36/*10+*36,*10+*36/*10+*36',\n",
       " '*2/*10,*10+*36/*14,*2x2/*10',\n",
       " '*2/*10+*36,*10+*36/*63,*2/*36x2',\n",
       " '*10+*36x2/*10+*36x2,*10x4+*36/*10+*36,*10+*36/*10+*36x2',\n",
       " '*10+*36/*10+*36x2,*10x2+*36/*10+*36,*10+*36/*10+*36',\n",
       " '*10/*10,*10/*10+*36,*10+*36/*10+*36',\n",
       " '*1/*10+*36x2,*1x2/*36x2,no_call',\n",
       " '*10/*10,*10/*10+*36,*10/*10x2+*36',\n",
       " '*2/*10+*36,*2/*10+*83,*2/*10',\n",
       " '*10+*36/*10+*36x2,*10x2+*36/*10+*36,*10+*36/*10+*36',\n",
       " '*5/*10+*36,*10/*36,*10/*10',\n",
       " '*2/*10,*2/*10+*36,*2/*10+*36x2',\n",
       " '*10/*10+*36,*10+*52/*10+*36+*68,*10x2/*52',\n",
       " '*10+*36x2/*41,*10+*36/*36+*41,*39/*69x2',\n",
       " '*10+*36x2/*132,*10+*36/*36+*132,*10/*10x3',\n",
       " '*10/*10+*36,*10+*36/*36,*10/*36',\n",
       " '*1/*10+*36,*1/*10,no_call',\n",
       " '*10/*10+*36,*10/*36,*10/*36x2',\n",
       " '*1/*10x2,*1/*10+*36,no_call',\n",
       " '*2/*10+*36x2,*2+*36/*10+*36,*10+*36x2/*39',\n",
       " '*13/*34,*1/*2,*1/*2x2',\n",
       " '*10+*36/*14,*2+*10+*36/*14+*68,*2/*10+*36',\n",
       " '*1/*10,*1/*10+*36,*1/*10+*36x2',\n",
       " '*10+*36/*41,*36+*39/*69,no_call',\n",
       " '*1/*10,*1/*10+*36,*1/*10x2+*36',\n",
       " '*10+*36/*41,*13/*36,no_call',\n",
       " '*1/*10,*1/*36,no_call',\n",
       " '*10/*10+*36,*10+*36/*10+*36,*10+*36/*10+*36x2',\n",
       " '*1/*41,*13/*63,*34/*119',\n",
       " '*10+*36/*10+*36x2,*10x2+*36/*10+*36,*10+*36/*10+*36',\n",
       " '*10/*10+*36,*10+*36/*10+*36,*10+*36/*10+*36x2',\n",
       " '*5/*10+*36,*10+*36/*68,*10/*36',\n",
       " '*2/*5,*1/*2,*1/*2x2',\n",
       " '*35/*49,*39/*49,no_call',\n",
       " '*2/*75,*1/*2,*1/*34',\n",
       " '*5/*41,*41/*41,*41/*41x2',\n",
       " '*2/*10+*36,*2/*36+*39,*2x2/*10',\n",
       " '*1/*10+*36,*1x2/*10+*36+*68,*1/*10x2',\n",
       " '*10+*36/*10+*36,*10x2+*36/*10+*36,*10/*10x3',\n",
       " '*5/*10+*36,*36/*36,*10/*36',\n",
       " '*10+*36/*41,*10+*36/*119,*39/*69x2',\n",
       " '*5/*10+*36,*10/*36,*10/*10',\n",
       " '*10/*10,*10/*10+*36,*10/*10x3',\n",
       " '*10/*10,*10/*10+*36,*10/*10x2+*36',\n",
       " '*39/*69,*10/*41,*36/*41',\n",
       " '*10+*36/*10+*36x2,*10x2+*36/*10+*36,*10+*36/*10+*36',\n",
       " '*1/*41,*39/*41,*2/*119',\n",
       " '*1/*10,*13/*36,*1/*10x2+*36',\n",
       " '*1/*2,*1/*63,*1/*34',\n",
       " '*10/*10+*36,*10+*36/*36,*10/*36x2',\n",
       " '*1/*2,*1+*63/*2,*1/*34',\n",
       " '*10+*36/*10+*36,*10x2+*36/*10+*36,*10/*10+*36',\n",
       " '*4/*5,*1/*4,*1/*4x2',\n",
       " '*1/*36,*1/*1,no_call',\n",
       " '*1/*10,*36/*39,*1/*36',\n",
       " '*10/*10,*10/*10+*36,*10/*10x3',\n",
       " '*10/*41,*10+*36/*41,*10+*36x2/*41',\n",
       " '*10/*10,*10/*10+*36,*10+*36/*10+*36',\n",
       " '*10+*36/*10+*36,*10x2+*36/*10+*36,*10/*10+*36',\n",
       " '*10/*10,*10/*10+*36,*10/*10x2+*36',\n",
       " '*2/*10,*2/*10+*36,*2/*10x2+*36',\n",
       " '*10/*10+*36,*10+*36/*79,*10/*10x2',\n",
       " '*10+*36/*10+*36,*10x2+*36/*10+*36,*10/*10x2',\n",
       " '*10/*10+*36,*10+*36/*10+*36,*10+*36/*10+*36x2',\n",
       " '*5/*10+*36,*36/*36,*10/*36',\n",
       " '*10+*36/*10+*36x2,*10x2+*36/*10+*36,*10/*10x3',\n",
       " '*10/*10,*10/*10+*36,*10/*10x2+*36',\n",
       " '*10/*10+*36,*10+*36/*36,*10/*36x2',\n",
       " '*4+*68/*10,*4/*5,*4x2/*10',\n",
       " '*1/*35,*35/*86,no_call',\n",
       " '*1/*10,*1/*10+*36,*1/*10x2+*36',\n",
       " '*1/*13,*1/*2,*1/*2x2',\n",
       " '*1/*14,*1/*63,no_call',\n",
       " '*10/*10+*36,*10+*36/*10+*36,*10+*36/*10x2+*36',\n",
       " '*2/*10+*36,*2/*10x2,*2x2/*10',\n",
       " '*10/*10+*36,*10+*36/*10+*36,*10+*36/*10+*36x2',\n",
       " '*10/*10+*36,*10+*36/*36,*10/*36x2',\n",
       " '*10+*36/*10+*36,*10x2+*36/*10+*36,*10/*10+*36',\n",
       " '*2/*10+*36x2,*2+*36/*10+*36,*36x2/*39',\n",
       " '*5/*10+*36,*10/*61,*10/*36',\n",
       " '*1/*41,*41+*83/*67,*34/*119',\n",
       " '*1/*10,*1/*10+*36,*1/*10+*36x2',\n",
       " '*10/*10,*10/*10+*36,*10+*36/*10+*36',\n",
       " '*10+*36/*10+*36,*10/*10+*36,*10/*10x3',\n",
       " '*1/*10+*36,*36x2/*39,*1/*36x2',\n",
       " '*10+*36/*10+*36x2,*10x2+*36/*10+*36,*10+*36/*10x2+*36',\n",
       " '*5/*35,*1/*35,*1/*35x2',\n",
       " '*2/*10,*2/*10+*36,*2/*10x2+*36',\n",
       " '*1/*41,*41/*83,no_call',\n",
       " '*10/*41,*10+*36/*41,*10x2+*36/*41',\n",
       " '*2/*10+*36x2,*2+*36/*10+*36,*10+*36/*39',\n",
       " '*10+*36x2/*21,*10+*36/*21+*36,*2/*10+*36',\n",
       " '*5/*10+*36,*10/*10+*36,*10+*36/*10+*36',\n",
       " '*5/*10+*36,*36/*36,*10/*36',\n",
       " '*2/*10+*36,*10+*68/*10+*14+*36,*10+*36/*14',\n",
       " '*2/*10,*2/*10+*36,*2/*10+*36x2',\n",
       " '*1/*10,*1/*10+*36,*1/*10+*36x2',\n",
       " '*2/*4+*68,no_call,*2/*4',\n",
       " '*1/*5,*1/*1,*1/*1x2',\n",
       " '*1/*10+*36x2,*1+*36/*10+*36,*1/*10x3',\n",
       " '*1x2/*10+*36,*1+*79/*36,no_call',\n",
       " '*10+*36/*10+*36,*10+*36/*36,*10/*10+*36',\n",
       " '*21/*36,*2/*36,*2/*10',\n",
       " '*5/*10+*36,*36/*36,*10/*36',\n",
       " '*10x2/*10+*36,*10+*36/*10+*36,*10/*36x2',\n",
       " '*1/*10+*36,*1+*36/*10+*36,*1/*10x2+*36',\n",
       " '*2/*10+*36x2,*10+*36/*83,*36x2/*39',\n",
       " '*1/*10+*36x2,*1+*36/*10+*36,*1/*10+*36',\n",
       " '*21/*31,*2/*21,*2/*2',\n",
       " '*1/*10+*36x2,*1+*36/*10+*36,no_call',\n",
       " '*10/*10+*36x2+*83,*10x2/*10+*36+*83,*10+*36x2/*39',\n",
       " '*2/*36x2,*10+*36/*63,*2/*10',\n",
       " '*5/*10+*36,*36/*36,no_call',\n",
       " '*10+*36x2/*49,*10+*36/*36+*49,*36x2/*49x2',\n",
       " '*1/*10,*1/*10+*36,*1/*10x2+*36',\n",
       " '*5/*10+*36,*36/*36,*10/*36',\n",
       " '*2/*10+*36,*2+*79/*36,*2/*10',\n",
       " '*10+*36x2/*10+*36x2,*10x3+*36/*10+*36,no_call',\n",
       " '*10+*36/*10+*36,*10x2+*36/*10+*36,*10/*10+*36',\n",
       " '*2/*10,*2/*10+*36,*2/*10x2+*36',\n",
       " '*10+*36/*10+*36x2,*10x2+*36/*10+*36,*10/*10x3',\n",
       " '*1/*43,*43/*86,no_call',\n",
       " '*10+*36/*41,*10+*36+*68/*41+*65,*2/*10+*36',\n",
       " '*2/*10+*36,*10x2/*63,*10+*36/*39',\n",
       " '*1/*10,*1/*10+*36,*1/*10x2+*36',\n",
       " '*10+*36/*10+*36x2,*10x2+*36/*10+*36,*10/*10x3',\n",
       " '*1/*10+*36,*1+*39/*36,no_call',\n",
       " '*1/*2,*13/*63,*1/*34',\n",
       " '*5/*41,*1/*41,*1/*41x2',\n",
       " '*10/*10,*10/*36,no_call',\n",
       " '*10+*36/*10+*36x2,*10x2+*36/*10+*36,*10/*10x3',\n",
       " '*2/*10+*36x2,*2+*36/*10+*36,*2/*10+*36',\n",
       " '*5/*5,*1/*5,no_call',\n",
       " '*2/*14,*14/*86,no_call',\n",
       " '*10/*10,*10/*10+*36,*10/*10x3',\n",
       " '*1/*10+*36,*1+*10/*10+*36+*68,no_call',\n",
       " '*5/*10+*36,*10/*10,*10/*36',\n",
       " '*10/*113,*10+*36/*113,*10x2+*36/*113',\n",
       " '*10/*10+*36,*10x2/*10+*36,*10/*10x2']"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Cyrius call\n",
    "no_consensus_df[no_consensus_df[\"cyrius\"] != \"no_call\"][[\"cyrius\", \"aldy\", \"stellarpgx\"]].to_csv(index=False).split()"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "b5b955d98a984774b329488fa6ec1e96bc983552e1b0cf177cdc63f1afde6ddf"
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
